File: genui/07_harness_genui/step_03_trueforge_hybrid/openui_parse.py
"""Generative UI step 03 - the step 02 OpenUI Lang parser, plus partial().

The language is a list of statements, one per line:

    root = Stack([header, kpis], "column", "l")
    header = TextContent("Lemonade stand", "large-heavy")
    kpis = Stack([Card([TextContent("" + total, "large-heavy")])], "row")
    total = 370

Values are strings, numbers, true/false/null, lists, inline component calls,
`+` (string concatenation), and identifiers that refer to other statements.
A reference may point forward to a statement that has not arrived yet; the
tree marks it as pending and fills it in when the line lands. That is what
makes the format stream: a page can draw the skeleton from the first line.

Parser.feed(text) takes any chunk of text; complete statements are parsed at
once, an unfinished line (open bracket, open string, no newline) is held
back. Parser.tree() resolves from "root" into plain Python: components are
{"type": name, "args": [...]}, pending references are {"type": "Pending",
"ref": id}.

New in step 03: Parser.partial() reads the held-back line leniently, so a
long statement such as `art = HtmlArtifact("Title", "<html>...` appears in
the tree as {"type": "HtmlArtifact", "args": [...], "partial": True} while
its document string is still arriving. tree() uses it for a reference whose
statement is the line in progress; everything else is step 02's parser.
"""

import re

TOKEN = re.compile(
    r'\s*(?:(?P<str>"(?:[^"\\]|\\.)*")|(?P<num>-?\d+(?:\.\d+)?)|(?P<id>[A-Za-z_]\w*)|(?P<punct>[()\[\],+=]))'
)
STRING_ESCAPES = {"n": "\n", "t": "\t", '"': '"', "\\": "\\"}
PARTIAL_HEAD = re.compile(r"\s*([A-Za-z_]\w*)\s*=\s*([A-Za-z_]\w*)\((.*)$", re.DOTALL)
SCALAR = re.compile(r"-?\d+(?:\.\d+)?|[A-Za-z_]\w*")


class ParseError(ValueError):
    pass


def read_string(text, start):
    """Decode the double-quoted string at `start`: (value, index after it, closed?).

    An unterminated string decodes as far as the text goes, a trailing lone
    backslash dropped, so a streaming document can be shown part-way.
    """
    out, i = [], start + 1
    while i < len(text):
        ch = text[i]
        if ch == "\\":
            if i + 1 >= len(text):
                break
            out.append(STRING_ESCAPES.get(text[i + 1], text[i + 1]))
            i += 2
        elif ch == '"':
            return "".join(out), i + 1, True
        else:
            out.append(ch)
            i += 1
    return "".join(out), len(text), False


def tokenize(line):
    """A statement line to a list of (kind, value) tokens; kinds: str num id punct."""
    tokens, pos = [], 0
    line = line.strip()
    while pos < len(line):
        m = TOKEN.match(line, pos)
        if not m or m.end() == pos:
            raise ParseError(f"unexpected text at {pos}: {line[pos:pos + 12]!r}")
        pos = m.end()
        kind = m.lastgroup
        raw = m.group(kind)
        if kind == "str":
            raw = re.sub(r"\\(.)", lambda e: STRING_ESCAPES.get(e.group(1), e.group(1)), raw[1:-1])
        elif kind == "num":
            raw = float(raw) if "." in raw else int(raw)
        tokens.append((kind, raw))
    return tokens


def complete(line):
    """True when brackets balance and no string is open: the line can be parsed."""
    depth, in_str, i = 0, False, 0
    while i < len(line):
        ch = line[i]
        if in_str:
            if ch == "\\":
                i += 1
            elif ch == '"':
                in_str = False
        elif ch == '"':
            in_str = True
        elif ch in "([":
            depth += 1
        elif ch in ")]":
            depth -= 1
        i += 1
    return depth <= 0 and not in_str


class Statement:
    """`name = expr`, with the expression kept as a small AST."""

    def __init__(self, name, expr):
        self.name = name
        self.expr = expr


def parse_statement(line):
    """One line to a Statement. Blank lines and # comments return None."""
    if not line.strip() or line.lstrip().startswith("#"):
        return None
    tokens = tokenize(line)
    if len(tokens) < 3 or tokens[0][0] != "id" or tokens[1] != ("punct", "="):
        raise ParseError(f"expected `name = value`: {line.strip()!r}")
    expr, end = parse_expr(tokens, 2)
    if end != len(tokens):
        raise ParseError(f"trailing tokens in {line.strip()!r}")
    return Statement(tokens[0][1], expr)


def parse_expr(tokens, i):
    """An expression: a value, or values joined by +. Returns (ast, next index)."""
    left, i = parse_value(tokens, i)
    while i < len(tokens) and tokens[i] == ("punct", "+"):
        right, i = parse_value(tokens, i + 1)
        left = ("add", left, right)
    return left, i


def parse_value(tokens, i):
    if i >= len(tokens):
        raise ParseError("unexpected end of statement")
    kind, value = tokens[i]
    if kind == "str":
        return ("str", value), i + 1
    if kind == "num":
        return ("num", value), i + 1
    if kind == "id":
        if value in ("true", "false"):
            return ("bool", value == "true"), i + 1
        if value == "null":
            return ("null", None), i + 1
        if i + 1 < len(tokens) and tokens[i + 1] == ("punct", "("):
            args, i = parse_list(tokens, i + 2, ")")
            return ("call", value, args), i
        return ("ref", value), i + 1
    if (kind, value) == ("punct", "["):
        items, i = parse_list(tokens, i + 1, "]")
        return ("list", items), i
    raise ParseError(f"unexpected token {value!r}")


def parse_list(tokens, i, closer):
    """Comma-separated expressions up to `closer`. Returns (items, next index)."""
    items = []
    while True:
        if i >= len(tokens):
            raise ParseError(f"missing {closer}")
        if tokens[i] == ("punct", closer):
            return items, i + 1
        item, i = parse_expr(tokens, i)
        items.append(item)
        if i < len(tokens) and tokens[i] == ("punct", ","):
            i += 1


class Parser:
    """Feed text in any pieces; read the resolved tree at any time."""

    def __init__(self):
        self.statements = {}   # name -> Statement, in arrival order
        self.buffer = ""       # text not yet ending in a newline, or unbalanced
        self.errors = []

    def feed(self, text):
        """Add a chunk. Every complete statement it finishes is parsed now."""
        self.buffer += text
        while True:
            # the first newline at which the text before it is balanced ends a statement
            start, cut = 0, None
            while cut is None:
                nl = self.buffer.find("\n", start)
                if nl == -1:
                    return  # the rest is an unfinished line: hold it back
                if complete(self.buffer[:nl]):
                    cut = nl
                start = nl + 1
            line, self.buffer = self.buffer[:cut], self.buffer[cut + 1:]
            self.add_line(line.replace("\n", " "))

    def partial(self):
        """The line in progress, read leniently: (name, type, args) or None.

        Only the shape `name = Component(scalar, scalar, ...)` is read, which
        is what a streaming HtmlArtifact looks like. Closed strings are
        decoded; the open one at the end is decoded as far as it goes.
        """
        head = PARTIAL_HEAD.match(self.buffer)
        if not head:
            return None
        name, kind, rest = head.group(1), head.group(2), head.group(3)
        args, i = [], 0
        while i < len(rest):
            ch = rest[i]
            if ch in " ,\n":
                i += 1
            elif ch == '"':
                value, i, closed = read_string(rest, i)
                args.append(value)
                if not closed:
                    break
            else:
                m = SCALAR.match(rest, i)
                if not m:
                    break
                raw = m.group(0)
                args.append(float(raw) if "." in raw else int(raw) if raw.lstrip("-").isdigit() else raw)
                i = m.end()
        return name, kind, args

    def close(self):
        """No more text is coming: parse whatever is still buffered."""
        if self.buffer.strip():
            self.add_line(self.buffer)
        self.buffer = ""

    def add_line(self, line):
        try:
            statement = parse_statement(line)
        except ParseError as e:
            self.errors.append(str(e))
            return
        if statement is not None:
            self.statements[statement.name] = statement

    # ----------------------------------------------------------- resolve

    def tree(self, name="root"):
        """The value behind `name`, with references followed and components as dicts."""
        if name not in self.statements:
            return {"type": "Pending", "ref": name}
        return self.resolve(self.statements[name].expr, {name})

    def resolve(self, expr, path):
        kind = expr[0]
        if kind in ("str", "num", "bool", "null"):
            return expr[1]
        if kind == "list":
            return [self.resolve(item, path) for item in expr[1]]
        if kind == "add":
            left, right = self.resolve(expr[1], path), self.resolve(expr[2], path)
            if isinstance(left, str) or isinstance(right, str):
                return f"{left}{right}"
            return left + right
        if kind == "call":
            return {"type": expr[1], "args": [self.resolve(a, path) for a in expr[2]]}
        name = expr[1]
        if name in path:
            return {"type": "Cycle", "ref": name}
        if name not in self.statements:
            partial = self.partial()  # the line in progress, if it is this statement
            if partial is not None and partial[0] == name:
                return {"type": partial[1], "args": partial[2], "partial": True}
            return {"type": "Pending", "ref": name}
        return self.resolve(self.statements[name].expr, path | {name})

    def pending(self):
        """Every referenced name that has no statement yet."""
        missing = []

        def walk(expr):
            kind = expr[0]
            if kind == "ref" and expr[1] not in self.statements and expr[1] not in missing:
                missing.append(expr[1])
            elif kind == "list":
                [walk(e) for e in expr[1]]
            elif kind == "call":
                [walk(e) for e in expr[2]]
            elif kind == "add":
                walk(expr[1]), walk(expr[2])

        for statement in self.statements.values():
            walk(statement.expr)
        return missing


def parse(program):
    """One-shot: the whole program to a Parser with every line consumed."""
    parser = Parser()
    parser.feed(program if program.endswith("\n") else program + "\n")
    parser.close()
    return parser


def outline(node, depth=0):
    """A component tree as indented lines, for the terminal."""
    pad = "  " * depth
    if isinstance(node, dict):
        if node["type"] in ("Pending", "Cycle"):
            return [f"{pad}({node['type'].lower()}: {node['ref']})"]
        scalars = [repr(a) if isinstance(a, str) else str(a) for a in node["args"] if not isinstance(a, (dict, list))]
        lines = [f"{pad}{node['type']}({', '.join(scalars)})"]
        for arg in node["args"]:
            if isinstance(arg, (dict, list)):
                lines += outline(arg, depth + 1)
        return lines
    if isinstance(node, list):
        if all(not isinstance(item, (dict, list)) for item in node):
            return [f"{pad}{node}"]
        return [line for item in node for line in outline(item, depth)]
    return [f"{pad}{node!r}"]

File: genui/07_harness_genui/step_03_trueforge_hybrid/web/openui-parse.mjs
// Generative UI step 03 - the OpenUI Lang parser, same shape as openui_parse.py.
//
// Statements `name = expr`, one per line. Values: "strings", numbers,
// true/false/null, [lists], Component(args), a + b, and references to other
// statements, forward or backward. feed() takes any chunk; tree() resolves
// from "root" with missing statements marked {type: "Pending", ref}.
//
// New in step 03: partial() reads the held-back line leniently, so a long
// statement such as `art = HtmlArtifact("Title", "<html>...` appears in the
// tree as {type: "HtmlArtifact", args, partial: true} while its document
// string is still arriving.

const TOKEN = /\s*(?:("(?:[^"\\]|\\.)*")|(-?\d+(?:\.\d+)?)|([A-Za-z_]\w*)|([()\[\],+=]))/y;
const ESCAPES = { n: "\n", t: "\t", '"': '"', "\\": "\\" };
const PARTIAL_HEAD = /^\s*([A-Za-z_]\w*)\s*=\s*([A-Za-z_]\w*)\(([\s\S]*)$/;
const SCALAR = /-?\d+(?:\.\d+)?|[A-Za-z_]\w*/y;

// Decode the double-quoted string at `start`: [value, index after it, closed].
// An unterminated string decodes as far as the text goes, a trailing lone
// backslash dropped, so a streaming document can be shown part-way.
export function readString(text, start) {
  let out = "";
  let i = start + 1;
  while (i < text.length) {
    const ch = text[i];
    if (ch === "\\") {
      if (i + 1 >= text.length) break;
      out += ESCAPES[text[i + 1]] ?? text[i + 1];
      i += 2;
    } else if (ch === '"') {
      return [out, i + 1, true];
    } else {
      out += ch;
      i++;
    }
  }
  return [out, text.length, false];
}

export function tokenize(line) {
  const tokens = [];
  line = line.trim();
  TOKEN.lastIndex = 0;
  while (TOKEN.lastIndex < line.length) {
    const start = TOKEN.lastIndex;
    const m = TOKEN.exec(line);
    if (!m || m.index !== start || m[0].length === 0) throw new Error(`unexpected text at ${start}: ${line.slice(start, start + 12)}`);
    if (m[1] !== undefined) tokens.push({ kind: "str", value: m[1].slice(1, -1).replace(/\\(.)/g, (_, c) => ESCAPES[c] ?? c) });
    else if (m[2] !== undefined) tokens.push({ kind: "num", value: Number(m[2]) });
    else if (m[3] !== undefined) tokens.push({ kind: "id", value: m[3] });
    else tokens.push({ kind: "punct", value: m[4] });
  }
  return tokens;
}

// True when brackets balance and no string is open: the line can be parsed.
export function complete(line) {
  let depth = 0;
  let inStr = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (inStr) {
      if (ch === "\\") i++;
      else if (ch === '"') inStr = false;
    } else if (ch === '"') inStr = true;
    else if (ch === "(" || ch === "[") depth++;
    else if (ch === ")" || ch === "]") depth--;
  }
  return depth <= 0 && !inStr;
}

export function parseStatement(line) {
  if (!line.trim() || line.trimStart().startsWith("#")) return null;
  const tokens = tokenize(line);
  if (tokens.length < 3 || tokens[0].kind !== "id" || tokens[1].value !== "=") throw new Error(`expected \`name = value\`: ${line.trim()}`);
  const [expr, end] = parseExpr(tokens, 2);
  if (end !== tokens.length) throw new Error(`trailing tokens in ${line.trim()}`);
  return { name: tokens[0].value, expr };
}

function parseExpr(tokens, i) {
  let [left, next] = parseValue(tokens, i);
  while (next < tokens.length && tokens[next].value === "+" && tokens[next].kind === "punct") {
    const [right, after] = parseValue(tokens, next + 1);
    left = ["add", left, right];
    next = after;
  }
  return [left, next];
}

function parseValue(tokens, i) {
  const token = tokens[i];
  if (!token) throw new Error("unexpected end of statement");
  if (token.kind === "str") return [["str", token.value], i + 1];
  if (token.kind === "num") return [["num", token.value], i + 1];
  if (token.kind === "id") {
    if (token.value === "true" || token.value === "false") return [["bool", token.value === "true"], i + 1];
    if (token.value === "null") return [["null", null], i + 1];
    if (tokens[i + 1] && tokens[i + 1].value === "(" && tokens[i + 1].kind === "punct") {
      const [args, next] = parseList(tokens, i + 2, ")");
      return [["call", token.value, args], next];
    }
    return [["ref", token.value], i + 1];
  }
  if (token.kind === "punct" && token.value === "[") {
    const [items, next] = parseList(tokens, i + 1, "]");
    return [["list", items], next];
  }
  throw new Error(`unexpected token ${token.value}`);
}

function parseList(tokens, i, closer) {
  const items = [];
  for (;;) {
    if (i >= tokens.length) throw new Error(`missing ${closer}`);
    if (tokens[i].kind === "punct" && tokens[i].value === closer) return [items, i + 1];
    const [item, next] = parseExpr(tokens, i);
    items.push(item);
    i = next;
    if (tokens[i] && tokens[i].kind === "punct" && tokens[i].value === ",") i++;
  }
}

export class Parser {
  constructor() {
    this.statements = new Map();
    this.buffer = "";
    this.errors = [];
  }

  // Add a chunk. Every complete statement it finishes is parsed now; an
  // unfinished line (open bracket, open string, no newline) is held back.
  feed(text) {
    this.buffer += text;
    for (;;) {
      let start = 0;
      let cut = -1;
      while (cut === -1) {
        const nl = this.buffer.indexOf("\n", start);
        if (nl === -1) return;
        if (complete(this.buffer.slice(0, nl))) cut = nl;
        start = nl + 1;
      }
      const line = this.buffer.slice(0, cut);
      this.buffer = this.buffer.slice(cut + 1);
      this.addLine(line.replace(/\n/g, " "));
    }
  }

  // The line in progress, read leniently: {name, type, args} or null. Only
  // the shape `name = Component(scalar, scalar, ...)` is read, which is what
  // a streaming HtmlArtifact looks like; the open string at the end is
  // decoded as far as it goes.
  partial() {
    const head = PARTIAL_HEAD.exec(this.buffer);
    if (!head) return null;
    const [, name, type, rest] = head;
    const args = [];
    let i = 0;
    while (i < rest.length) {
      const ch = rest[i];
      if (ch === " " || ch === "," || ch === "\n") i++;
      else if (ch === '"') {
        const [value, next, closed] = readString(rest, i);
        args.push(value);
        i = next;
        if (!closed) break;
      } else {
        SCALAR.lastIndex = i;
        const m = SCALAR.exec(rest);
        if (!m) break;
        args.push(/^-?\d/.test(m[0]) ? Number(m[0]) : m[0]);
        i = SCALAR.lastIndex;
      }
    }
    return { name, type, args };
  }

  close() {
    if (this.buffer.trim()) this.addLine(this.buffer);
    this.buffer = "";
  }

  addLine(line) {
    try {
      const statement = parseStatement(line);
      if (statement) this.statements.set(statement.name, statement.expr);
    } catch (e) {
      this.errors.push(e.message);
    }
  }

  tree(name = "root") {
    if (!this.statements.has(name)) return { type: "Pending", ref: name };
    return this.resolve(this.statements.get(name), new Set([name]));
  }

  resolve(expr, path) {
    const [kind] = expr;
    if (kind === "str" || kind === "num" || kind === "bool" || kind === "null") return expr[1];
    if (kind === "list") return expr[1].map((item) => this.resolve(item, path));
    if (kind === "add") {
      const left = this.resolve(expr[1], path);
      const right = this.resolve(expr[2], path);
      return typeof left === "string" || typeof right === "string" ? `${left}${right}` : left + right;
    }
    if (kind === "call") return { type: expr[1], args: expr[2].map((a) => this.resolve(a, path)) };
    const name = expr[1];
    if (path.has(name)) return { type: "Cycle", ref: name };
    if (!this.statements.has(name)) {
      const partial = this.partial(); // the line in progress, if it is this statement
      if (partial && partial.name === name) return { type: partial.type, args: partial.args, partial: true };
      return { type: "Pending", ref: name };
    }
    return this.resolve(this.statements.get(name), new Set([...path, name]));
  }

  // Every referenced name that has no statement yet.
  pending() {
    const missing = [];
    const walk = (expr) => {
      const [kind] = expr;
      if (kind === "ref" && !this.statements.has(expr[1]) && !missing.includes(expr[1])) missing.push(expr[1]);
      else if (kind === "list") expr[1].forEach(walk);
      else if (kind === "call") expr[2].forEach(walk);
      else if (kind === "add") { walk(expr[1]); walk(expr[2]); }
    };
    for (const expr of this.statements.values()) walk(expr);
    return missing;
  }
}

export function parse(program) {
  const parser = new Parser();
  parser.feed(program.endsWith("\n") ? program : program + "\n");
  parser.close();
  return parser;
}
